File: services/crown-adapter/src/events.ts
import { randomUUID } from "node:crypto";

import { PRODUCER_VERSION, SCHEMA_VERSION } from "@intent/contracts";

import { CROWN_CHANNELS } from "./quality.ts";
import { monotonicNs } from "./timestamps.ts";

export const SOURCE = "crown-adapter";

export type UnnormalizedEvent = {
  schema_version: string;
  event_id: string;
  event_type: string;
  source: string;
  modality: string | null;
  session_id: string | null;
  trial_id: string | null;
  sequence: number;
  source_time_ns: number | null;
  received_monotonic_ns: number;
  quality: number;
  producer_version: string;
  payload: Record<string, unknown>;
};

export function makeEvent(opts: {
  eventType: string;
  modality: string | null;
  sequence: number;
  sourceTimeNs: number | null;
  quality?: number;
  payload: Record<string, unknown>;
}): UnnormalizedEvent {
  return {
    schema_version: SCHEMA_VERSION,
    event_id: randomUUID().replaceAll("-", ""),
    event_type: opts.eventType,
    source: SOURCE,
    modality: opts.modality,
    session_id: null,
    trial_id: null,
    sequence: opts.sequence,
    source_time_ns: opts.sourceTimeNs,
    received_monotonic_ns: monotonicNs(),
    quality: opts.quality ?? 1,
    producer_version: PRODUCER_VERSION,
    payload: opts.payload,
  };
}

export function heartbeatEvent(
  sequence: number,
  uptimeSeconds: number,
  lastDataAgeMs: number | null,
  errorCount: number,
  status = "healthy",
): UnnormalizedEvent {
  return makeEvent({
    eventType: "service.heartbeat",
    modality: null,
    sequence,
    sourceTimeNs: null,
    payload: {
      status,
      uptime_seconds: uptimeSeconds,
      last_data_age_ms: lastDataAgeMs,
      error_count: errorCount,
    },
  });
}

export function deviceStatusEvent(
  sequence: number,
  status: "healthy" | "degraded" | "offline",
  detail: string | null,
  metadata: Record<string, unknown> = {},
): UnnormalizedEvent {
  return makeEvent({
    eventType: "device.status",
    modality: "eeg",
    sequence,
    sourceTimeNs: null,
    payload: {
      status,
      device_alias: "crown-mock",
      detail,
      battery_percent: 92,
      metadata: {
        stream: "raw",
        os_version: "mock",
        ...metadata,
      },
    },
  });
}

export { CROWN_CHANNELS };

File: services/crown-adapter/crown_adapter/events.py
from __future__ import annotations

from typing import Any

from intent_contracts.enums import PRODUCER_VERSION, SCHEMA_VERSION
from intent_contracts.envelope import EventEnvelope, new_event_id, now_monotonic_ns
from intent_runtime.heartbeat import heartbeat_event as runtime_heartbeat

from crown_adapter.quality import CROWN_CHANNELS

SOURCE = "crown-adapter"

__all__ = ["CROWN_CHANNELS", "SOURCE", "device_status", "heartbeat", "make_event"]


def make_event(
    *,
    event_type: str,
    sequence: int,
    payload: dict[str, Any],
    modality: str | None = "eeg",
    source_time_ns: int | None = None,
    quality: float = 1.0,
    received_monotonic_ns: int | None = None,
) -> EventEnvelope:
    return EventEnvelope(
        schema_version=SCHEMA_VERSION,
        event_id=new_event_id(),
        event_type=event_type,
        source=SOURCE,
        modality=modality,
        session_id=None,
        trial_id=None,
        sequence=sequence,
        source_time_ns=source_time_ns,
        received_monotonic_ns=(
            received_monotonic_ns if received_monotonic_ns is not None else now_monotonic_ns()
        ),
        quality=quality,
        producer_version=PRODUCER_VERSION,
        payload=payload,
    )


def heartbeat(
    sequence: int,
    uptime_seconds: float,
    last_data_age_ms: float | None,
    error_count: int,
    status: str = "healthy",
) -> EventEnvelope:
    return runtime_heartbeat(
        SOURCE,
        uptime_seconds=uptime_seconds,
        last_data_age_ms=last_data_age_ms,
        error_count=error_count,
        sequence=sequence,
        status=status,
    )


def device_status(
    sequence: int,
    status: str,
    detail: str | None,
    metadata: dict[str, Any] | None = None,
    device_alias: str = "crown-mock",
) -> EventEnvelope:
    payload: dict[str, Any] = {
        "status": status,
        "device_alias": device_alias,
        "detail": detail,
        "metadata": {"stream": "raw", **(metadata or {})},
    }
    if device_alias == "crown-mock":
        payload["battery_percent"] = 92
        payload["metadata"]["os_version"] = "mock"
    return make_event(
        event_type="device.status",
        sequence=sequence,
        payload=payload,
        quality=0.0 if status == "offline" else 1.0,
    )
